core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -517,47 +526,90 @@ class TransactionCoordinator(txnConfig: TransactionConfig,
-                  // We should clear the pending state to make way for the transition to PrepareAbort and also bump
-                  // the epoch in the transaction metadata we are about to append.
-                  isEpochFence = true
-                  txnMetadata.pendingState = None
-                  txnMetadata.producerEpoch = producerEpoch
-                  txnMetadata.lastProducerEpoch = RecordBatch.NO_PRODUCER_EPOCH
-                }
+                // We should only be in retrying states on Prepare/Complete states.
+                // If we are in ongoing, some other producer must have transitioned that state.
+                if (retryOnOverflow || retryOnEpochBump) {
+                  Left(Errors.PRODUCER_FENCED)
+                } else {
+                  val nextState = if (txnMarkerResult == TransactionResult.COMMIT)
+                    PrepareCommit
+                  else
+                    PrepareAbort
+
+                  // Maybe allocate new producer ID if we are bumping epoch and epoch is exhausted
+                  val nextProducerIdOrErrors =
+                    if (requestIsAtLeastTransactionsV2 && !txnMetadata.pendingState.contains(PrepareEpochFence) && txnMetadata.isProducerEpochExhausted) {
+                      producerIdManager.generateProducerId() match {
+                        case Success(newProducerId) =>
+                          Right(newProducerId)
+                        case Failure(exception) =>
+                          Left(Errors.forException(exception))
+                      }
+                    } else {
+                      Right(RecordBatch.NO_PRODUCER_ID)
+                    }
+
+                  if (nextState == PrepareAbort && txnMetadata.pendingState.contains(PrepareEpochFence)) {
+                    // We should clear the pending state to make way for the transition to PrepareAbort and also bump
+                    // the epoch in the transaction metadata we are about to append.
+                    isEpochFence = true
+                    txnMetadata.pendingState = None
+                    txnMetadata.producerEpoch = producerEpoch
+                    txnMetadata.lastProducerEpoch = RecordBatch.NO_PRODUCER_EPOCH
+                  }
 
-                Right(coordinatorEpoch, txnMetadata.prepareAbortOrCommit(nextState, time.milliseconds()))
+                  nextProducerIdOrErrors match {
+                    case Left(error) =>
+                      Left(error)
+                    case Right(nextProducerId) =>
+                      Right(coordinatorEpoch, txnMetadata.prepareAbortOrCommit(nextState, clientTransactionVersion, nextProducerId, time.milliseconds()))
+                  }
+                }
               case CompleteCommit =>
-                if (txnMarkerResult == TransactionResult.COMMIT)
+                if (txnMarkerResult == TransactionResult.COMMIT && validCompleteTransition)
```

Comment:
It looks like `validCompleteTransition` is effectively epoch check and should result in producer fenced error, but it results in invalid transition error.

Reply:
Hmmm. I don't know if I agree. The producer is not fenced in this case. In this case, the epoch equals the txnMetadata and it is invalid. 

If the epoch is < the txn state, it will get caught in the fenced case above. I tried to explain in the comment, but maybe it is unclear.

Reply:
This is not consistent with what we had before: unexpected epoch led to producer fenced error:
```
if ((isFromClient && producerEpoch != txnMetadata.producerEpoch) || producerEpoch < txnMetadata.producerEpoch)
              Left(Errors.PRODUCER_FENCED)
```
If we decide to redefine this condition, we should redefine it consistently for all states.

Reply:
I don't think that's quite true because the epoch is expected when it is the same epoch -- it was just the wrong state. We just never had a case where same epoch == invalidState unless it was the wrong type of marker. 

I don't feel super strongly though since the result will be the same from the client side.

Reply:
I have updated with the producer fenced approach